In [1]:
import pandas as pd
import numpy as np

In [2]:
titanic=pd.read_csv("titanic.csv")
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked deck
0         0       3    male  22.0      1      0   7.2500        S  NaN
1         1       1  female  38.0      1      0  71.2833        C    C
2         1       3  female  26.0      0      0   7.9250        S  NaN
3         1       1  female  35.0      1      0  53.1000        S    C
4         0       3    male  35.0      0      0   8.0500        S  NaN

In [3]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  891 non-null    int64  
 1   pclass    891 non-null    int64  
 2   sex       891 non-null    object 
 3   age       714 non-null    float64
 4   sibsp     891 non-null    int64  
 5   parch     891 non-null    int64  
 6   fare      891 non-null    float64
 7   embarked  889 non-null    object 
 8   deck      203 non-null    object 
dtypes: float64(2), int64(4), object(3)
memory usage: 62.8+ KB


In [4]:
# We can create a data series by just selecting any column from the df

titanic["deck"]

0      NaN
1        C
2      NaN
3        C
4      NaN
      ... 
886    NaN
887      B
888    NaN
889      C
890    NaN
Name: deck, Length: 891, dtype: object

In [5]:
# Now we can check whether the data type was really a "series" data type

type(titanic["deck"])

pandas.core.series.Series

In [6]:
# Note: the head and tail method works for the series data as well

In [7]:
deck1=titanic["deck"]

In [8]:
deck1.head()

0    NaN
1      C
2    NaN
3      C
4    NaN
Name: deck, dtype: object

In [9]:
deck1.tail()

886    NaN
887      B
888    NaN
889      C
890    NaN
Name: deck, dtype: object

In [10]:
deck1.dtype

dtype('O')

In [11]:
deck1.shape # Its saying the series is a 1 dimensional array with 891 element

(891,)

In [12]:
len(deck1)

891

In [13]:
deck1.index

RangeIndex(start=0, stop=891, step=1)

In [14]:
deck1.describe()

count     203
unique      7
top         C
freq       59
Name: deck, dtype: object

## Analyzing Numerical Series

In [15]:
age1=titanic["age"]
age1.head()

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: age, dtype: float64

In [16]:
age1.dtype # This defines how bytes in memory will be interpreted by a scalar (i.e. a single number) or an array and the 
            # way in which the bytes will be treated (e.g. int/float).

dtype('float64')

In [17]:
type(age1) # this is just the type of Python object it is (similar to how type("hello") is str for example).

pandas.core.series.Series

In [18]:
# The count method returns all the non-null values in the data series

age1.count()

714

In [19]:
# Size attribute ,on the other hand, returns total number of all the available values in the series

age1.size

891

In [20]:
# Function length

len(age1)

891

### x.sum() vs sum(x)

x.sum() has a default setting of skipping over missing values
sum(x) does not do that. Hence, if there is na value in your series or df and you run sum(x), you will get nan as a result

### And not surprisingly, x.sum() comes from a panda library. Panda is designed to handle such real world 
### problems. Coz in the real world data, there are lots of missing values

In [21]:
age1.sum()

21205.17

In [22]:
sum(age1)

nan

In [23]:
# Unique method returns unique values from the series. 
# Note that this feature is only available for panda data series and not for panda dataframe


age1.unique() # This basically means although we have 891 passengers, these are the unique ages of all the passengers aboard

array([22.  , 38.  , 26.  , 35.  ,   nan, 54.  ,  2.  , 27.  , 14.  ,
        4.  , 58.  , 20.  , 39.  , 55.  , 31.  , 34.  , 15.  , 28.  ,
        8.  , 19.  , 40.  , 66.  , 42.  , 21.  , 18.  ,  3.  ,  7.  ,
       49.  , 29.  , 65.  , 28.5 ,  5.  , 11.  , 45.  , 17.  , 32.  ,
       16.  , 25.  ,  0.83, 30.  , 33.  , 23.  , 24.  , 46.  , 59.  ,
       71.  , 37.  , 47.  , 14.5 , 70.5 , 32.5 , 12.  ,  9.  , 36.5 ,
       51.  , 55.5 , 40.5 , 44.  ,  1.  , 61.  , 56.  , 50.  , 36.  ,
       45.5 , 20.5 , 62.  , 41.  , 52.  , 63.  , 23.5 ,  0.92, 43.  ,
       60.  , 10.  , 64.  , 13.  , 48.  ,  0.75, 53.  , 57.  , 80.  ,
       70.  , 24.5 ,  6.  ,  0.67, 30.5 ,  0.42, 34.5 , 74.  ])

In [24]:
# How many unique values do we have in our series?

len(age1.unique())

89

In [25]:
# Another way to count the unique values is to use:

age1.nunique() # We got 88 here instead of 89 because the nunique() method by default skips over the missing value. and we
                # know for a fact that there is 1 na value in the above result.
    
    # If we change the dropna to False, it would result in 89 as well

88

In [26]:
age1.nunique(dropna=False)

89

### Another very powerful method: value_counts()

In [27]:
age1.value_counts() # This basically returns object containing of uniqzúe values. The most frequently occuring object 
                    # appears first on the series. In our case, it is the age 24 which appears 30 times.

24.00    30
22.00    27
18.00    26
19.00    25
30.00    25
         ..
55.50     1
70.50     1
66.00     1
23.50     1
0.42      1
Name: age, Length: 88, dtype: int64

In [28]:
# Note that by default, the value_counts() ignores all the missing values i.e. by default it is dropna=True. But if we change it:

age1.value_counts(dropna=False)

# As you can see, if the missing values are included, then by far the most frequently occuring elements are indeed the
# missing values

NaN      177
24.00     30
22.00     27
18.00     26
28.00     25
        ... 
36.50      1
55.50      1
66.00      1
23.50      1
0.42       1
Name: age, Length: 89, dtype: int64

Another thing that is mentioned above is about sorting. By default, the resulting data is arranged from most frequently
occuring to the least frequently occuring whenever you use the value_counts() method.

Hoever, you can change this by simply calling False on the sort:

In [29]:
age1.value_counts(dropna=True, sort=False, ascending=False) # Not surprisingly, this results in an unsorted series

22.00    27
38.00    11
26.00    18
35.00    18
54.00     8
         ..
0.92      1
0.83      2
0.67      1
70.50     1
0.42      1
Name: age, Length: 88, dtype: int64

In [30]:
# We can also sort the result in a ascending manner i.e. it goes from smallest to biggest

age1.value_counts(dropna=True, sort=True, ascending=True)

0.42      1
23.50     1
66.00     1
70.50     1
55.50     1
         ..
30.00    25
19.00    25
18.00    26
22.00    27
24.00    30
Name: age, Length: 88, dtype: int64

 We can also normalize the results if we already know the total number of x. For example, in our case about age of the
 people in titanic, we can get a sorta weights of different age group by dividing the frequency of a certain age 
 by total number of passengers

In [31]:
age1.value_counts(dropna=True, sort=True, ascending=False, normalize=True)

# Interpretation: The age group 24 represents 4.20% of the total passengers in the titanic

24.00    0.042017
22.00    0.037815
18.00    0.036415
19.00    0.035014
30.00    0.035014
           ...   
55.50    0.001401
70.50    0.001401
66.00    0.001401
23.50    0.001401
0.42     0.001401
Name: age, Length: 88, dtype: float64

In [32]:
# USEAGE OF BINS

age1.value_counts(dropna=True, sort=True, ascending=False, bins=5).head()

(16.336, 32.252]    346
(32.252, 48.168]    188
(0.339, 16.336]     100
(48.168, 64.084]     69
(64.084, 80.0]       11
Name: age, dtype: int64

### Analyzing Non-numerical Series

In [33]:
import pandas as pd

In [34]:
summer=pd.read_csv("summer.csv")

In [35]:
summer.head()

Year    City     Sport Discipline             Athlete Country Gender  \
0  1896  Athens  Aquatics   Swimming       HAJOS, Alfred     HUN    Men   
1  1896  Athens  Aquatics   Swimming    HERSCHMANN, Otto     AUT    Men   
2  1896  Athens  Aquatics   Swimming   DRIVAS, Dimitrios     GRE    Men   
3  1896  Athens  Aquatics   Swimming  MALOKINIS, Ioannis     GRE    Men   
4  1896  Athens  Aquatics   Swimming  CHASAPIS, Spiridon     GRE    Men   

                        Event   Medal  
0              100M Freestyle    Gold  
1              100M Freestyle  Silver  
2  100M Freestyle For Sailors  Bronze  
3  100M Freestyle For Sailors    Gold  
4  100M Freestyle For Sailors  Silver

In [36]:
summer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31165 entries, 0 to 31164
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Year        31165 non-null  int64 
 1   City        31165 non-null  object
 2   Sport       31165 non-null  object
 3   Discipline  31165 non-null  object
 4   Athlete     31165 non-null  object
 5   Country     31161 non-null  object
 6   Gender      31165 non-null  object
 7   Event       31165 non-null  object
 8   Medal       31165 non-null  object
dtypes: int64(1), object(8)
memory usage: 2.1+ MB


In [37]:
athlete=summer["Athlete"]

In [38]:
type(athlete)

pandas.core.series.Series

In [39]:
athlete.head()

0         HAJOS, Alfred
1      HERSCHMANN, Otto
2     DRIVAS, Dimitrios
3    MALOKINIS, Ioannis
4    CHASAPIS, Spiridon
Name: Athlete, dtype: object

In [40]:
athlete.tail()

31160           JANIKOWSKI, Damian
31161    REZAEI, Ghasem Gholamreza
31162               TOTROV, Rustam
31163            ALEKSANYAN, Artur
31164               LIDBERG, Jimmy
Name: Athlete, dtype: object

In [41]:
athlete.dtype # This Oth object means that its a string object

dtype('O')

In [42]:
athlete.shape

(31165,)

In [43]:
athlete.describe()

count               31165
unique              22762
top       PHELPS, Michael
freq                   22
Name: Athlete, dtype: object

In [44]:
athlete.size # It counts all the values including NaN values

31165

In [45]:
athlete.count() # It skips over the na values. In our case, both give the same result because there are no missing values

31165

In [46]:
athlete.unique() # These are the people who won the medals in the summer olympics game

array(['HAJOS, Alfred', 'HERSCHMANN, Otto', 'DRIVAS, Dimitrios', ...,
       'TOTROV, Rustam', 'ALEKSANYAN, Artur', 'LIDBERG, Jimmy'],
      dtype=object)

In [47]:
# Q) But what is the number of people that won medals?

len(athlete.unique())

22762

## OR

In [48]:
athlete.nunique()

22762

## The Copy () method

In [49]:
titanic=pd.read_csv("titanic.csv")
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked deck
0         0       3    male  22.0      1      0   7.2500        S  NaN
1         1       1  female  38.0      1      0  71.2833        C    C
2         1       3  female  26.0      0      0   7.9250        S  NaN
3         1       1  female  35.0      1      0  53.1000        S    C
4         0       3    male  35.0      0      0   8.0500        S  NaN

In [50]:
age=titanic.age
age.head()

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: age, dtype: float64

Now lets imagine that we want to change the 3rd element (index pos 2) in this "age" slice that we have selected

In [51]:
age.iloc[2]=29
age.head()

C:\Users\user\Anaconda3\lib\site-packages\pandas\core\indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


0    22.0
1    38.0
2    29.0
3    35.0
4    35.0
Name: age, dtype: float64

And now if we call the original complete dataset, what do we see in index pos 2?

In [52]:
titanic.head() 

survived  pclass     sex   age  sibsp  parch     fare embarked deck
0         0       3    male  22.0      1      0   7.2500        S  NaN
1         1       1  female  38.0      1      0  71.2833        C    C
2         1       3  female  29.0      0      0   7.9250        S  NaN
3         1       1  female  35.0      1      0  53.1000        S    C
4         0       3    male  35.0      0      0   8.0500        S  NaN

As you can see, the element in the index pos 2 has changed from 26 to 29. This is not what we wanted. We just wanted to change the element in the age slice and not alter the original data

Is there any solution to this?

Yes, the solution is to use the copy() method when creating the first age slice itself

In [53]:
titanic=pd.read_csv("titanic.csv")
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked deck
0         0       3    male  22.0      1      0   7.2500        S  NaN
1         1       1  female  38.0      1      0  71.2833        C    C
2         1       3  female  26.0      0      0   7.9250        S  NaN
3         1       1  female  35.0      1      0  53.1000        S    C
4         0       3    male  35.0      0      0   8.0500        S  NaN

In [54]:
age=titanic["age"].copy()
age.head()

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: age, dtype: float64

In [55]:
age.iloc[2]=29
age.head()

0    22.0
1    38.0
2    29.0
3    35.0
4    35.0
Name: age, dtype: float64

In [56]:
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked deck
0         0       3    male  22.0      1      0   7.2500        S  NaN
1         1       1  female  38.0      1      0  71.2833        C    C
2         1       3  female  26.0      0      0   7.9250        S  NaN
3         1       1  female  35.0      1      0  53.1000        S    C
4         0       3    male  35.0      0      0   8.0500        S  NaN

So, as you can see above, the index pos 2 in the original dataset is still the same even though the age slice is altered

### Sorting a panda series

In [57]:
import pandas as pd

In [58]:
dic={1:10, 3:25, 2:6, 4:36, 5:2, 6:0, 7:None}
dic

{1: 10, 2: 6, 3: 25, 4: 36, 5: 2, 6: 0, 7: None}

In [59]:
sales=pd.Series(dic)
sales

1    10.0
3    25.0
2     6.0
4    36.0
5     2.0
6     0.0
7     NaN
dtype: float64

### Introducing sort_index()

In [60]:
sales.sort_index() # By default it is sorted in ascending way

1    10.0
2     6.0
3    25.0
4    36.0
5     2.0
6     0.0
7     NaN
dtype: float64

In [61]:
# But we can also do in descending order

sales.sort_index(ascending=False)

7     NaN
6     0.0
5     2.0
4    36.0
3    25.0
2     6.0
1    10.0
dtype: float64

Now, lets quickly check on our original data to see if anything has been altered.

In [62]:
sales

1    10.0
3    25.0
2     6.0
4    36.0
5     2.0
6     0.0
7     NaN
dtype: float64

As you can see above, nothing has changed. Well, this is because of a parameter inside of sort_index() called "inplace".
This "inplace" parameter when False, simply creates a copy of the original variable and hence, the original data is not 
altered.

However, if we do wanna change the original data, we can simply put the inplace parameter as True

In [63]:
sales.sort_index(ascending=False, inplace=True) # Note: ascending is by default False

In [64]:
sales

7     NaN
6     0.0
5     2.0
4    36.0
3    25.0
2     6.0
1    10.0
dtype: float64

As you can see above, the original data is still the same because of inplace 

#### SImilarly, we can also sort the value by using .sort_values()

In [65]:
sales

7     NaN
6     0.0
5     2.0
4    36.0
3    25.0
2     6.0
1    10.0
dtype: float64

In [66]:
sales.sort_values() # By default, ascending is True and inplace is False

6     0.0
5     2.0
2     6.0
1    10.0
3    25.0
4    36.0
7     NaN
dtype: float64

So, lets try some other variations

In [67]:
sales.sort_values(ascending=False, inplace=True)
sales

4    36.0
3    25.0
1    10.0
2     6.0
5     2.0
6     0.0
7     NaN
dtype: float64

We can also decide to put the na values at last or first (or whatever you want)

In [68]:
sales.sort_values(na_position="first")

7     NaN
6     0.0
5     2.0
2     6.0
1    10.0
3    25.0
4    36.0
dtype: float64

## Coding Exercise 3

### Exercise 3: Analyzing Columns / Pandas Series

In [69]:
import pandas as pd

In [70]:
import os
os.getcwd()
os.chdir("C:\\Users\\user\\Documents\\Udemy\\Python Finance\\Course_Materials_Part4\\Exercises")

In [71]:
listings=pd.read_csv("listings.csv")
listings.head()

Symbol Exchange                           Name  Last_Price    Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68  2.593470e+10   
1     AA     NYSE              Alcoa Corporation       29.15  5.407810e+09   
2   AABA   NASDAQ                    Altaba Inc.       75.39  4.278113e+10   
3    AAC     NYSE             AAC Holdings, Inc.        2.16  5.314109e+07   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02  1.527687e+10   

   ADR TSO  IPO_Year            Sector  \
0      NaN    1999.0     Capital Goods   
1      NaN    2016.0  Basic Industries   
2      NaN       NaN        Technology   
3      NaN    2014.0       Health Care   
4      NaN       NaN    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

In [72]:
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6745 non-null   float64
 4   Market_Cap  5954 non-null   float64
 5   ADR TSO     140 non-null    float64
 6   IPO_Year    3105 non-null   float64
 7   Sector      5309 non-null   object 
 8   Industry    5309 non-null   object 
dtypes: float64(4), object(5)
memory usage: 481.9+ KB


19. __Select__ the numerical column __"Last_Price"__, create a __copy__ and __save__ the column/Pandas Series in the variable __price__! __Fill in the gaps!__

In [73]:
price=listings["Last_Price"].copy()
price.head()

0    81.68
1    29.15
2    75.39
3     2.16
4    34.02
Name: Last_Price, dtype: float64

20. Get some __summary statistics__ on the Series __price__! What is the __price__ of the __most expensive stock__?

In [74]:
price.describe()

count    6745.000000
mean       37.372232
std        99.249040
min         0.003000
25%         8.310000
50%        20.573500
75%        37.370000
max      4299.990000
Name: Last_Price, dtype: float64

Hence, the price of most expensive stock is 4299.99

21. Get the __minimum Value__ in the Series __price__ by explicitly calling the ... method! The cheapest stock has a __Last_Price__ of...?

In [75]:
price.min()

0.003

The cheapest stock has the last price of 0.003

22. Get the Frequency/__Counts__ of all unique __values__ in the Series __price__! What is the __most frequent__ price?

In [76]:
price.value_counts()

9.7300      9
25.0000     9
1.6100      8
2.5000      8
10.2600     7
           ..
46.8000     1
131.3235    1
35.3000     1
53.9100     1
74.0000     1
Name: Last_Price, Length: 4470, dtype: int64

The most frequent price is 9.73 and 25

23. Get the __relative frequencies__ in the Series __price__! What is the __relative frequency__ of the __most frequent price__?

In [77]:
price.value_counts(normalize=True)

9.7300      0.001334
25.0000     0.001334
1.6100      0.001186
2.5000      0.001186
10.2600     0.001038
              ...   
46.8000     0.000148
131.3235    0.000148
35.3000     0.000148
53.9100     0.000148
74.0000     0.000148
Name: Last_Price, Length: 4470, dtype: float64

The relative frequency of most frequent price is 0.133%

24. __Sort__ the Series __price__ from __low to high__! What is the __second lowest__ price?

In [78]:
price.sort_values()

1676    0.003
3375    0.004
6439    0.005
6650    0.007
5684    0.010
        ...  
6170      NaN
6403      NaN
6551      NaN
6717      NaN
6839      NaN
Name: Last_Price, Length: 6852, dtype: float64

The second lowest price is 0.004

25. __Sort__ the Series __price__ from __high to low__ and __save the changes__ by setting the __inplace__ parameter to __True__!  <br>
__Fill in the gaps!__

In [79]:
price.sort_values(ascending=False, inplace=True)
price

5505    4299.9900
4500    2910.4200
350     1847.3300
815     1806.0000
631     1316.6999
          ...    
6170          NaN
6403          NaN
6551          NaN
6717          NaN
6839          NaN
Name: Last_Price, Length: 6852, dtype: float64

__Inspect__ the __first 5 elements__ of price! What is the __second highest__ price?

The second highest price is 2910.42

26. __Sort__ the Series __price__ by the __Index__ (ascending) and __save__ the changes!

In [80]:
price=price.sort_index(ascending=True)
price.head()

0    81.68
1    29.15
2    75.39
3     2.16
4    34.02
Name: Last_Price, dtype: float64

27. __Select__ the non-numerical column __"Sector"__, create a __copy__ and __save__ the column/Pandas Series in the variable __sector__! 

In [81]:
sector=listings["Sector"]
sector.head()

0       Capital Goods
1    Basic Industries
2          Technology
3         Health Care
4      Transportation
Name: Sector, dtype: object

Inspect! The first element/sector is ... ?

The first element is Capital Goods

28. Call the __describe()__ method on the non-numerical Series __"sector"__! What is the __most frequent__ value/sector?

In [82]:
sector.describe()

count        5309
unique         12
top       Finance
freq         1059
Name: Sector, dtype: object

The most frequent sector is Finance

29. Get all __unique values__ in the Series __sector__!

In [83]:
sector.unique()

array(['Capital Goods', 'Basic Industries', 'Technology', 'Health Care',
       'Transportation', 'Finance', 'Consumer Services', nan,
       'Consumer Durables', 'Consumer Non-Durables', 'Miscellaneous',
       'Public Utilities', 'Energy'], dtype=object)

30. Last but not least, __count the frequencies__ in the Series __sector__! How often does the value/sector __Transportation__ appear?

In [84]:
sector.value_counts("Transportation")

Finance                  0.199473
Health Care              0.169147
Consumer Services        0.155585
Technology               0.116594
Capital Goods            0.073272
Basic Industries         0.059145
Energy                   0.055001
Public Utilities         0.049727
Consumer Non-Durables    0.044453
Miscellaneous            0.028631
Consumer Durables        0.025805
Transportation           0.023168
Name: Sector, dtype: float64

The transportation sector appears 2.34% of the time

# Pandas Index Objects

## First steps

In [85]:
import pandas as pd

In [86]:
import os
os.chdir("C:\\Users\\user\\Documents\\Udemy\\Python Finance\\Course_Materials_Part4\\Video_Lectures_NBs")

In [87]:
summer=pd.read_csv("summer.csv", index_col="Athlete")
summer.head()

Year    City     Sport Discipline Country Gender  \
Athlete                                                                
HAJOS, Alfred       1896  Athens  Aquatics   Swimming     HUN    Men   
HERSCHMANN, Otto    1896  Athens  Aquatics   Swimming     AUT    Men   
DRIVAS, Dimitrios   1896  Athens  Aquatics   Swimming     GRE    Men   
MALOKINIS, Ioannis  1896  Athens  Aquatics   Swimming     GRE    Men   
CHASAPIS, Spiridon  1896  Athens  Aquatics   Swimming     GRE    Men   

                                         Event   Medal  
Athlete                                                 
HAJOS, Alfred                   100M Freestyle    Gold  
HERSCHMANN, Otto                100M Freestyle  Silver  
DRIVAS, Dimitrios   100M Freestyle For Sailors  Bronze  
MALOKINIS, Ioannis  100M Freestyle For Sailors    Gold  
CHASAPIS, Spiridon  100M Freestyle For Sailors  Silver

In [88]:
summer.index # The attribute index gives us info about the row index.


Index(['HAJOS, Alfred', 'HERSCHMANN, Otto', 'DRIVAS, Dimitrios',
       'MALOKINIS, Ioannis', 'CHASAPIS, Spiridon', 'CHOROPHAS, Efstathios',
       'HAJOS, Alfred', 'ANDREOU, Joannis', 'CHOROPHAS, Efstathios',
       'NEUMANN, Paul',
       ...
       'AHMADOV, Emin', 'KAZAKEVIC, Aleksandr', 'KHUGAEV, Alan',
       'EBRAHIM, Karam Mohamed Gaber', 'GAJIYEV, Danyal', 'JANIKOWSKI, Damian',
       'REZAEI, Ghasem Gholamreza', 'TOTROV, Rustam', 'ALEKSANYAN, Artur',
       'LIDBERG, Jimmy'],
      dtype='object', name='Athlete', length=31165)

In [89]:
type(summer.index)

pandas.core.indexes.base.Index

In [90]:
# We can do the same for columns

summer.columns

Index(['Year', 'City', 'Sport', 'Discipline', 'Country', 'Gender', 'Event',
       'Medal'],
      dtype='object')

In [91]:
type(summer.columns)

pandas.core.indexes.base.Index

In [92]:
# If you wanna look at both x and y axes,

summer.axes

[Index(['HAJOS, Alfred', 'HERSCHMANN, Otto', 'DRIVAS, Dimitrios',
        'MALOKINIS, Ioannis', 'CHASAPIS, Spiridon', 'CHOROPHAS, Efstathios',
        'HAJOS, Alfred', 'ANDREOU, Joannis', 'CHOROPHAS, Efstathios',
        'NEUMANN, Paul',
        ...
        'AHMADOV, Emin', 'KAZAKEVIC, Aleksandr', 'KHUGAEV, Alan',
        'EBRAHIM, Karam Mohamed Gaber', 'GAJIYEV, Danyal', 'JANIKOWSKI, Damian',
        'REZAEI, Ghasem Gholamreza', 'TOTROV, Rustam', 'ALEKSANYAN, Artur',
        'LIDBERG, Jimmy'],
       dtype='object', name='Athlete', length=31165),
 Index(['Year', 'City', 'Sport', 'Discipline', 'Country', 'Gender', 'Event',
        'Medal'],
       dtype='object')]

In [93]:
summer.columns[:3] # This gives us all the columns starting from pos 0 to pos 2 since pos 3 is exclusive

Index(['Year', 'City', 'Sport'], dtype='object')

In [94]:
summer.index[0]

'HAJOS, Alfred'

In [95]:
# SImilarly, you can also select the last statement by simply:

summer.index[-1]

'LIDBERG, Jimmy'

In [96]:
summer.columns.tolist() # This simply converts the index attribute to a list

['Year', 'City', 'Sport', 'Discipline', 'Country', 'Gender', 'Event', 'Medal']

In [97]:
summer.index.is_unique

False

In [98]:
summer.index.get_loc("DRIVAS, Dimitrios") # Tell us which index pos does this player lie on

2

## Changing Row Index Labels

In [99]:
import pandas as pd

In [100]:
summer=pd.read_csv("summer.csv")

In [101]:
summer.head()

Year    City     Sport Discipline             Athlete Country Gender  \
0  1896  Athens  Aquatics   Swimming       HAJOS, Alfred     HUN    Men   
1  1896  Athens  Aquatics   Swimming    HERSCHMANN, Otto     AUT    Men   
2  1896  Athens  Aquatics   Swimming   DRIVAS, Dimitrios     GRE    Men   
3  1896  Athens  Aquatics   Swimming  MALOKINIS, Ioannis     GRE    Men   
4  1896  Athens  Aquatics   Swimming  CHASAPIS, Spiridon     GRE    Men   

                        Event   Medal  
0              100M Freestyle    Gold  
1              100M Freestyle  Silver  
2  100M Freestyle For Sailors  Bronze  
3  100M Freestyle For Sailors    Gold  
4  100M Freestyle For Sailors  Silver

Lets check the index in this df

In [102]:
summer.index

RangeIndex(start=0, stop=31165, step=1)

### Now lets change the index. 

In [103]:
summer=pd.read_csv("summer.csv", index_col="Athlete")
summer.head()

Year    City     Sport Discipline Country Gender  \
Athlete                                                                
HAJOS, Alfred       1896  Athens  Aquatics   Swimming     HUN    Men   
HERSCHMANN, Otto    1896  Athens  Aquatics   Swimming     AUT    Men   
DRIVAS, Dimitrios   1896  Athens  Aquatics   Swimming     GRE    Men   
MALOKINIS, Ioannis  1896  Athens  Aquatics   Swimming     GRE    Men   
CHASAPIS, Spiridon  1896  Athens  Aquatics   Swimming     GRE    Men   

                                         Event   Medal  
Athlete                                                 
HAJOS, Alfred                   100M Freestyle    Gold  
HERSCHMANN, Otto                100M Freestyle  Silver  
DRIVAS, Dimitrios   100M Freestyle For Sailors  Bronze  
MALOKINIS, Ioannis  100M Freestyle For Sailors    Gold  
CHASAPIS, Spiridon  100M Freestyle For Sailors  Silver

Lets again check the index 

In [104]:
summer.index

Index(['HAJOS, Alfred', 'HERSCHMANN, Otto', 'DRIVAS, Dimitrios',
       'MALOKINIS, Ioannis', 'CHASAPIS, Spiridon', 'CHOROPHAS, Efstathios',
       'HAJOS, Alfred', 'ANDREOU, Joannis', 'CHOROPHAS, Efstathios',
       'NEUMANN, Paul',
       ...
       'AHMADOV, Emin', 'KAZAKEVIC, Aleksandr', 'KHUGAEV, Alan',
       'EBRAHIM, Karam Mohamed Gaber', 'GAJIYEV, Danyal', 'JANIKOWSKI, Damian',
       'REZAEI, Ghasem Gholamreza', 'TOTROV, Rustam', 'ALEKSANYAN, Artur',
       'LIDBERG, Jimmy'],
      dtype='object', name='Athlete', length=31165)

### But what if we are not happy and we wanna change this back to the "range" index?

In [105]:
summer.reset_index(inplace=True) # Note the parameter "drop" in the reset_index() is by default set to False. If it is set to True, then
                    # the column that is being replaced by the range index col is dropped/deleted from the df
    
                    # Same goes from the "inplace" parameter. By default, it is False. But if you wanna change the orginal
                    # data, go ahead and set the parameter to True

In [106]:
summer.head()

Athlete  Year    City     Sport Discipline Country Gender  \
0       HAJOS, Alfred  1896  Athens  Aquatics   Swimming     HUN    Men   
1    HERSCHMANN, Otto  1896  Athens  Aquatics   Swimming     AUT    Men   
2   DRIVAS, Dimitrios  1896  Athens  Aquatics   Swimming     GRE    Men   
3  MALOKINIS, Ioannis  1896  Athens  Aquatics   Swimming     GRE    Men   
4  CHASAPIS, Spiridon  1896  Athens  Aquatics   Swimming     GRE    Men   

                        Event   Medal  
0              100M Freestyle    Gold  
1              100M Freestyle  Silver  
2  100M Freestyle For Sailors  Bronze  
3  100M Freestyle For Sailors    Gold  
4  100M Freestyle For Sailors  Silver

### Now, lets say we wanna have the "year" column as index?!

There are two ways to do this: 1) while loading the data which we already did above 2) we can use the .set_index() method

In [107]:
summer.set_index("Year", inplace=True)

In [108]:
summer.head()

Athlete    City     Sport Discipline Country Gender  \
Year                                                                   
1896       HAJOS, Alfred  Athens  Aquatics   Swimming     HUN    Men   
1896    HERSCHMANN, Otto  Athens  Aquatics   Swimming     AUT    Men   
1896   DRIVAS, Dimitrios  Athens  Aquatics   Swimming     GRE    Men   
1896  MALOKINIS, Ioannis  Athens  Aquatics   Swimming     GRE    Men   
1896  CHASAPIS, Spiridon  Athens  Aquatics   Swimming     GRE    Men   

                           Event   Medal  
Year                                      
1896              100M Freestyle    Gold  
1896              100M Freestyle  Silver  
1896  100M Freestyle For Sailors  Bronze  
1896  100M Freestyle For Sailors    Gold  
1896  100M Freestyle For Sailors  Silver

In [109]:
summer.index.is_unique # It is basically checking whether the elements in ´the index column are all unique or have some 
                        # duplicate components in it

False

In [110]:
summer.reset_index()

Year                    Athlete    City      Sport  \
0      1896              HAJOS, Alfred  Athens   Aquatics   
1      1896           HERSCHMANN, Otto  Athens   Aquatics   
2      1896          DRIVAS, Dimitrios  Athens   Aquatics   
3      1896         MALOKINIS, Ioannis  Athens   Aquatics   
4      1896         CHASAPIS, Spiridon  Athens   Aquatics   
...     ...                        ...     ...        ...   
31160  2012         JANIKOWSKI, Damian  London  Wrestling   
31161  2012  REZAEI, Ghasem Gholamreza  London  Wrestling   
31162  2012             TOTROV, Rustam  London  Wrestling   
31163  2012          ALEKSANYAN, Artur  London  Wrestling   
31164  2012             LIDBERG, Jimmy  London  Wrestling   

                Discipline Country Gender                       Event   Medal  
0                 Swimming     HUN    Men              100M Freestyle    Gold  
1                 Swimming     AUT    Men              100M Freestyle  Silver  
2                 Swimming     GRE    Men  100M Freestyle For Sailors  Bronze  
3                 Swimming     GRE    Men  100M Freestyle For Sailors    Gold  
4                 Swimming     GRE    Men  100M Freestyle For Sailors  Silver  
...                    ...     ...    ...                         ...     ...  
31160  Wrestling Freestyle     POL    Men                    Wg 84 KG  Bronze  
31161  Wrestling Freestyle     IRI    Men                    Wg 96 KG    Gold  
31162  Wrestling Freestyle     RUS    Men                    Wg 96 KG  Silver  
31163  Wrestling Freestyle     ARM    Men                    Wg 96 KG  Bronze  
31164  Wrestling Freestyle     SWE    Men                    Wg 96 KG  Bronze  

[31165 rows x 9 columns]

## Changing Column Labels

In [111]:
titanic= pd.read_csv("titanic.csv")
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked deck
0         0       3    male  22.0      1      0   7.2500        S  NaN
1         1       1  female  38.0      1      0  71.2833        C    C
2         1       3  female  26.0      0      0   7.9250        S  NaN
3         1       1  female  35.0      1      0  53.1000        S    C
4         0       3    male  35.0      0      0   8.0500        S  NaN

In [112]:
titanic.tail()

survived  pclass     sex   age  sibsp  parch   fare embarked deck
886         0       2    male  27.0      0      0  13.00        S  NaN
887         1       1  female  19.0      0      0  30.00        S    B
888         0       3  female   NaN      1      2  23.45        S  NaN
889         1       1    male  26.0      0      0  30.00        C    C
890         0       3    male  32.0      0      0   7.75        Q  NaN

In [113]:
titanic.columns # Checking columnn labels with the attributes column

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'deck'],
      dtype='object')

In [114]:
titanic.columns[0]

'survived'

#### What if we wanna change a particular column name?

In [115]:
titanic.columns[0]="Alive"

# However, we get an error saying that the index does not support mutable operations. This simply means that you cannot 
# change just one column name. If you wanna change, you gotta change every column name

TypeError: Index does not support mutable operations

For example, if we do this, it will probably work:

In [ ]:
titanic.columns=["Alive","Class","Sex","Age","SibSp","ParChi","Fare","Emb","Deck"]
titanic.head()

In [ ]:
# How to change or assign name for out column index ?

In [ ]:
titanic.columns.name # When we run this, nothing shows up. This means there is no name

In [ ]:
titanic.columns.name="Pass_characteristics"
titanic.head()

#### But what about the name of row index?

In [ ]:
titanic.index.name="Passenger_no"

In [ ]:
titanic.head()

### Renaming indexing (row) and column labels by using the .rename() method. 
#### There are two ways to do this: 1) To use the .rename(mapper={}) 2) To use the .rename(index={})

In [ ]:
summer=pd.read_csv("summer.csv", index_col="Athlete")
summer.head()

In [ ]:
# Now CHANGING AN ELEMENT FROM THE INDEX COLUMN

summer.rename(mapper={"HAJOS, Alfred":'HAYOS, Alfred'}, axis="index")

In [ ]:
summer.rename(index={"HAJOS, Alfred":'HAYOS, Alfred'}, inplace=True) # inplace True coz we really wanted to change the 
                                                                    # original data

In [ ]:
summer.head()

### We can do the same for column labels

In [ ]:
summer.rename(mapper={"City":"Location"}, axis="columns")

# OR

In [ ]:
summer.rename(columns={"Sport":"Category of Sport"}, inplace =True)

In [ ]:
summer.head()

### Exercise 4: Index Operations

In [ ]:
import os

In [ ]:
os.getcwd()

In [ ]:
os.chdir("C:\\Users\\user\\Documents\\Udemy\\Python Finance\\Course_Materials_Part4\\Exercises")

In [ ]:
listings=pd.read_csv("listings.csv")
listings.head()

31. __Select__ the __index__ (row labels) of the DataFrame __listings__! What __kind of index__ do we have?

In [ ]:
listings.index

We have a range index

32. __Set__ the __"Symbol"__ column as the __new index__ of listings (save the changes)!

In [ ]:
listings.set_index("Symbol", inplace=True)

In [ ]:
listings.head()

33. __Check__ whether the new index contains __unique values only__ (no duplicates)! Is this the case?

In [ ]:
listings.index.is_unique

34. __Get__ the __frequency/Counts__ of all unique values in the index! What is the __most frequent__ value(s)?

In [ ]:
listings.index.value_counts()

35. __Set__ the __name__ of the index to __"Ticker_Symbol"__!

In [ ]:
listings.index.name

In [ ]:
listings.index.name="Ticker_Symbol"

In [ ]:
listings.head()

37. __Inspect__ the __column Index__!

In [ ]:
listings.columns.name

Above code gives out blank. This means there is no column index name 

38. __Rename__ the column __Labels__ "Name" and "Market_Cap" to "Company" and "MCAP"! __Fill in  the gaps__!

In [ ]:
listings.rename(columns={"Name":"Company","Market_Cap":"MCAP"}, inplace=True)

In [ ]:
listings.head()

## Sorting dfs with sort_index() and sort_values

In [ ]:
import pandas as pd

In [ ]:
import os
os.getcwd()

In [ ]:
os.chdir("C:\\Users\\user\\Documents\\Udemy\\Python Finance\\Course_Materials_Part4\\Video_Lectures_NBs")

In [ ]:
titanic=pd.read_csv("titanic.csv")

In [ ]:
titanic.head()

In [ ]:
titanic.tail()

In [ ]:
titanic.age.sort_values(ascending=False)

In [ ]:
titanic.head() 

# The orginal data is still the same because by default the inplace parameter in the sort_values() method is set False

#### ANother method doing the same as above:

In [ ]:
titanic.sort_values("age", axis=0, ascending =False, inplace=False)

#### We can also perform many operations at once:

In [ ]:
titanic.sort_values(["pclass","sex","age"], ascending=False) 

# This code really provides the sorting for the mentioned columns one after the other. It means first the pclass is sorted
# in a descending manner and then, the sex column and finally, the age column.

#### What if you wanted to have one column descending and the other column ascending?

In [ ]:
titanic.sort_values(["pclass","sex","age"], ascending=["False","True","False"]) 

## nunique(), nlargest() and nsmallest() with df

In [ ]:
import pandas as pd

In [ ]:
titanic=pd.read_csv("titanic.csv")
titanic.head()

In [ ]:
titanic.tail()

In [ ]:
titanic.age.unique()

#### But if we wanna have unique data for the entire df, we do this:

In [ ]:
titanic.nunique(axis=0) 

# So here, we have unique values for different columns in our df
# This means in the survived column, we have 2 unique values and so on

If we wanna include the na elements that are in df:

In [ ]:
titanic.nunique(axis=0, dropna=False)

In [ ]:
titanic.nlargest(n=5, columns="fare")

#### Another way to do the same as above is by using .sort_values():

In [ ]:
titanic.sort_values("fare",ascending=False).head()

#### Who are the 5 most youngest passengers on board?

In [ ]:
titanic.nsmallest(n=5,columns="age")

#### Can you say who is the youngest passenger on board?

In [ ]:
titanic.nsmallest(n=1, columns="age")

#### Another method to calculate the same thing?

In [ ]:
titanic.loc[titanic.age.idxmin()]

## Filtering df with one condition

In [ ]:
import pandas as pd

In [ ]:
titanic=pd.read_csv("titanic.csv")

In [ ]:
titanic.head(10)

#### Now, what if we want to create a df with only male ?

In [ ]:
# Lets first observe our first 10 rows of sex column 

titanic.sex.head(10)

#### wanna check the above result?

In [ ]:
titanic.sex=="male" # This results in panda series not a df

#### Now, if we pass the command for above panda series with boolean values into a sq.bracket of the titanic dataset, we get:

In [ ]:
titanic[titanic.sex=="male"]


# Now, as you can see in the sex column, there are only male data

#### What if you wanna have data of all the male and how much fare did they pay?

#### You can do that by using chain indexing

In [ ]:
titanic[titanic.sex=="male"]["fare"]

#### There is also actually an another way to extract the same output: using loc

In [ ]:
titanic.loc[titanic.sex=="male","fare"]

#### PRO TIP: Avoid chain indexing as much as possible. It creates big problem later in data manipulation

#### We can also assign the variable mask to the boolean series

In [ ]:
mask1=titanic.sex=="male"
mask1

#### And to further increase the readability of the code:

In [ ]:
titanic_male=titanic.loc[mask1]
titanic_male.head()

#### With loc notation, we can filter for both columns and rows. Also for data types

In [ ]:
titanic.dtypes

#### Now, suppose you wanna have a df with only string objects. This means now you need column which have the dtype of string objects 

In [ ]:
mask2=titanic.dtypes==object
mask2

In [ ]:
titanic.loc[:, ~mask2]

#### Lets add 1 more layer of restriction to that. Now we want all the male whose column belongs to the data type strings and all the columns of datatypes except for strings

In [ ]:
titanic.loc[mask1, ~mask2]

#### Filtering df with many conditions (AND)

In [ ]:
import pandas as pd

In [ ]:
titanic=pd.read_csv("titanic.csv")
titanic.head()

#### Condition 1

In [ ]:
mask1=titanic.sex=="male"
mask1.head()

#### Condition 2

In [ ]:
mask2=titanic.age>14
mask2

#### Essentially, what we are asking here is to generate a new series or df which has all the men who are above the age of 14

#### The solution for that is use the AND operator. Since we already have two boolean dfs, it should work pretty well in filtering out the data.

In [ ]:
(mask1 & mask2).head() # Remember that it only results in "True" if both it is True in both the series

#### So, the survived male are:

In [ ]:
male_surv=titanic.loc[mask1 & mask2]
male_surv.head()

#### However, we only want these columns: survived, sex, pclass and age

In [ ]:
male_surv=titanic.loc[mask1 & mask2, ["survived","pclass","sex","age"]]
male_surv.head()

#### Can you give some descriptive statistics on this newly formed df?

In [ ]:
male_surv.describe()

#### We can compare the above stats with the stats of the original dataset

In [ ]:
titanic.describe()

#### So, from observing the the full dataset and the newly formed df, we can come up with some good analytical points.

#### The total surviving poulation % was 38.38. However, among the males who were above the age of 14, the survival rate decreases drastically to just 17.39. This really shows that being an adult male on the ship during the disaster period was a bad characteristic to have.

### Filtering the df with many conditions (OR)

#### We want to answer this question: Was being a child or a woman advantageous in the sinking ship?

In [ ]:
titanic.head()

In [ ]:
filter1=titanic.sex=="female"
filter1.head()

In [ ]:
filter2=titanic.age < 14
filter2.head()

In [ ]:
(filter1 | filter2).head()

In [ ]:
survived_womchi=titanic.loc[(filter1|filter2),["survived","age","pclass","sex"]]
survived_womchi.head()

In [ ]:
survived_womchi.describe()

#### As we can clearly see, the survival rate of women or children has gone up to 72.36%. This is way way above men's survival rate of 17 % and still way above the survival rate of the overall sample which had the rate of 38 %.

### Advanced filtering with between(), isin(), and ~

In [ ]:
import pandas as pd

In [ ]:
summer=pd.read_csv("summer.csv")

In [ ]:
summer.head()

In [ ]:
medals_1988=summer.loc[summer.Year==1988]
medals_1988.head()

In [ ]:
medals_1988.info()

#### What if we wanna have the data of the medals won since 1992?

In [ ]:
medals_since92=summer.loc[summer.Year>=1992]
medals_since92.head()

In [ ]:
medals_since92.tail()

In [ ]:
medals_since92.info()

#### Now, lets check whether we could find all the years for olympic games played in the decade of 1960s 

In [ ]:
summer.Year.between(1960,1969).head()

#### We get a boolean series of False in the first few rows. This is because our original summer dataset starts from 1896.
#### Now we can pass the above boolean series to get a new df

In [ ]:
medals_in60s=summer.loc[summer.Year.between(1960,1969, inclusive=True)]
medals_in60s.head()

In [ ]:
medals_in60s.tail()

#### Now, imagine your favorite olympic games were the year of 2012 and 1996. How would you create a df with only these two years?

In [ ]:
my_fav_years=[2012,1996]

In [ ]:
summer.Year.isin(my_fav_years).head()

In [ ]:
medals_12and96=summer.loc[summer.Year.isin(my_fav_years)]
medals_12and96.head()

#### Just to flip around the above idea on its head, and ask what to do if we want the data of all the years except 2012 and 1996?

In [ ]:
medals_except_12and96=summer.loc[~summer.Year.isin(my_fav_years)]
medals_except_12and96.head()

In [ ]:
medals_except_12and96.tail()

#### But, how do we check?

In [ ]:
medals_except_12and96.Year.unique()

## Learn how to use the methods: any() and all()

In [ ]:
import pandas as pd

In [ ]:
titanic=pd.read_csv("titanic.csv")
titanic.head()

#### Creating a boolean series to check the passengers are male 

In [ ]:
titanic.sex=="male"

#### You wanna check if any one of the element in the boolean series above happens to be "male". What do you do?

In [ ]:
(titanic.sex=="male").any()

#### We get "True" because in the boolean series above, we will definitely find "at least" one male 

#### Similarly, how to check if all passengers were male?

In [ ]:
(titanic.sex=="male").all()

#### And not surprisingly, we get False because obviously titanic was not filled with only males

#### Similarly, we can also take this concept of any and all and apply in finding a bit more details about the dataset

In [ ]:
(titanic.age==60.0).any()

#### The fact that we got True from the above command means that there is definitely at least one person on the titanic who was 60 years old

### Did you know that the boolean value "True" represents not just 1 rather all the values that are not 0 and "False" is always only 0

### Exercise 5: Sorting & Filtering 

In [ ]:
import pandas as pd
import os

In [ ]:
os.getcwd()
os.chdir("C:\\Users\\user\\Documents\\Udemy\\Python Finance\\Course_Materials_Part4\\Exercises")

In [ ]:
listings=pd.read_csv("listings.csv")
listings.head()

39. __Sort__ the __listings__ DataFrame by the __"Market_Cap"__ column from __high to low__ and __save__ the new order! __Fill in the gaps__!

In [ ]:
listings.sort_values("Market_Cap", ascending=False, inplace=True)
listings

__Inspect__ listings! Which Company has the __highest Market Capitalization__ (Name)?

#### The company with the highest market cap is Apple

40. __Sort__ listings by __IPO_Year__ (ascending) and __Market_Cap__ (descending) and save the new order. 

In [ ]:
listings.head()

In [ ]:
listings.sort_values(["IPO_Year","Market_Cap"], ascending=[True, False], inplace=True)
listings.head()

#### In the above solution, you can see that the IPO years are sorted in an ascending manner. However, the market cap is not really descending even though we wrote the correct code for it.

#### Well the reason is because the code is for making the market cap sort in descending after the IPO rankings. This means that once the IPO year has been sorted in ascendingly then, the companies which have same IPO year are in take the same rank. Hence, the descending market cap focuses on this sub-section and arranges the companies with the same ipo year on the basis of their market cap from high to low aka descendingly

__Inspect__ listings! From all Companies that were listed in __1983__, which one has the __highest Market Capitalization__?

In [116]:
listings.head()

Symbol Exchange                           Name  Last_Price    Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68  2.593470e+10   
1     AA     NYSE              Alcoa Corporation       29.15  5.407810e+09   
2   AABA   NASDAQ                    Altaba Inc.       75.39  4.278113e+10   
3    AAC     NYSE             AAC Holdings, Inc.        2.16  5.314109e+07   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02  1.527687e+10   

   ADR TSO  IPO_Year            Sector  \
0      NaN    1999.0     Capital Goods   
1      NaN    2016.0  Basic Industries   
2      NaN       NaN        Technology   
3      NaN    2014.0       Health Care   
4      NaN       NaN    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

41. Set the column __Symbol__ as new Index of the __listings__ DataFrame! __Save__ the change!

In [117]:
listings.index

RangeIndex(start=0, stop=6852, step=1)

In [119]:
listings.set_index("Symbol", inplace=True)
listings.head()

Exchange                           Name  Last_Price    Market_Cap  \
Symbol                                                                     
A          NYSE     Agilent Technologies, Inc.       81.68  2.593470e+10   
AA         NYSE              Alcoa Corporation       29.15  5.407810e+09   
AABA     NASDAQ                    Altaba Inc.       75.39  4.278113e+10   
AAC        NYSE             AAC Holdings, Inc.        2.16  5.314109e+07   
AAL      NASDAQ  American Airlines Group, Inc.       34.02  1.527687e+10   

        ADR TSO  IPO_Year            Sector  \
Symbol                                        
A           NaN    1999.0     Capital Goods   
AA          NaN    2016.0  Basic Industries   
AABA        NaN       NaN        Technology   
AAC         NaN    2014.0       Health Care   
AAL         NaN       NaN    Transportation   

                                                Industry  
Symbol                                                    
A       Biotechnology: Laboratory Analytical Instruments  
AA                                              Aluminum  
AABA                                        EDP Services  
AAC                                 Medical Specialities  
AAL                        Air Freight/Delivery Services

42. __Sort__ the __listings__ DataFrame by the new __index__ from __low to high__. __Save__ the new order!

In [121]:
listings.sort_index(ascending=True, inplace=True)
listings.head()

Exchange                           Name  Last_Price    Market_Cap  \
Symbol                                                                     
A          NYSE     Agilent Technologies, Inc.       81.68  2.593470e+10   
AA         NYSE              Alcoa Corporation       29.15  5.407810e+09   
AABA     NASDAQ                    Altaba Inc.       75.39  4.278113e+10   
AAC        NYSE             AAC Holdings, Inc.        2.16  5.314109e+07   
AAL      NASDAQ  American Airlines Group, Inc.       34.02  1.527687e+10   

        ADR TSO  IPO_Year            Sector  \
Symbol                                        
A           NaN    1999.0     Capital Goods   
AA          NaN    2016.0  Basic Industries   
AABA        NaN       NaN        Technology   
AAC         NaN    2014.0       Health Care   
AAL         NaN       NaN    Transportation   

                                                Industry  
Symbol                                                    
A       Biotechnology: Laboratory Analytical Instruments  
AA                                              Aluminum  
AABA                                        EDP Services  
AAC                                 Medical Specialities  
AAL                        Air Freight/Delivery Services

#### Going back to RangeIndex by reseting the current one

In [122]:
listings.reset_index(inplace=True)
listings.head()

Symbol Exchange                           Name  Last_Price    Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68  2.593470e+10   
1     AA     NYSE              Alcoa Corporation       29.15  5.407810e+09   
2   AABA   NASDAQ                    Altaba Inc.       75.39  4.278113e+10   
3    AAC     NYSE             AAC Holdings, Inc.        2.16  5.314109e+07   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02  1.527687e+10   

   ADR TSO  IPO_Year            Sector  \
0      NaN    1999.0     Capital Goods   
1      NaN    2016.0  Basic Industries   
2      NaN       NaN        Technology   
3      NaN    2014.0       Health Care   
4      NaN       NaN    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

43. Check the number of __unique values__ in __all columns__ of the __listings__ DataFrame! __How many__ unique values are in the __"Exchange"__ column?

In [124]:
listings.nunique()

Symbol        6840
Exchange         3
Name          5766
Last_Price    4470
Market_Cap    5835
ADR TSO        140
IPO_Year        42
Sector          12
Industry       135
dtype: int64

#### There are 3 unique values in the Exchange column

44. __Select__ the __10__ listings with the __highest Market_Cap__! What is the Company __Name__ of the __third largest__ listing?

In [130]:
listings.nlargest(n=10, columns="Market_Cap")

Symbol Exchange                           Name  Last_Price    Market_Cap  \
11     AAPL   NASDAQ                     Apple Inc.      200.62  9.459795e+11   
4109   MSFT   NASDAQ          Microsoft Corporation      120.19  9.221233e+11   
350    AMZN   NASDAQ               Amazon.com, Inc.     1847.33  9.074138e+11   
2767  GOOGL   NASDAQ                  Alphabet Inc.     1206.45  8.387076e+11   
2766   GOOG   NASDAQ                  Alphabet Inc.     1202.16  8.357253e+11   
2262     FB   NASDAQ                 Facebook, Inc.      177.82  5.074965e+11   
623    BABA     NYSE  Alibaba Group Holding Limited      186.19  4.826388e+11   
3434    JNJ     NYSE              Johnson & Johnson      135.58  3.610186e+11   
6412      V     NYSE                      Visa Inc.      158.56  3.590806e+11   
6770    XOM     NYSE        Exxon Mobil Corporation       81.56  3.453905e+11   

      ADR TSO  IPO_Year             Sector  \
11        NaN    1980.0         Technology   
4109      NaN    1986.0         Technology   
350       NaN    1997.0  Consumer Services   
2767      NaN       NaN         Technology   
2766      NaN    2004.0         Technology   
2262      NaN    2012.0         Technology   
623       NaN    2014.0      Miscellaneous   
3434      NaN       NaN        Health Care   
6412      NaN       NaN      Miscellaneous   
6770      NaN       NaN             Energy   

                                             Industry  
11                             Computer Manufacturing  
4109          Computer Software: Prepackaged Software  
350                    Catalog/Specialty Distribution  
2767  Computer Software: Programming, Data Processing  
2766  Computer Software: Programming, Data Processing  
2262  Computer Software: Programming, Data Processing  
623                                 Business Services  
3434                            Major Pharmaceuticals  
6412                                Business Services  
6770                         Integrated oil Companies

#### Therefore, the third largest Market cap in the data is AMZN

## OR 
#### You can also accomplish the same result like this:

In [136]:
listings.index

Int64Index([6851, 6850, 6849, 6848, 6847, 6846, 6845, 6844, 6843, 6842,
            ...
               9,    8,    7,    6,    5,    4,    3,    2,    1,    0],
           dtype='int64', length=6852)

In [137]:
listings.set_index("Market_Cap", inplace=True)
listings.head()

Symbol Exchange                           Name  Last_Price  \
Market_Cap                                                                
1.618349e+08   ZYXI   NASDAQ                    Zynex, Inc.        5.02   
1.653995e+08   ZYNE   NASDAQ  Zynerba Pharmaceuticals, Inc.        7.85   
5.040782e+08   ZYME     NYSE                 Zymeworks Inc.       15.74   
2.147266e+09    ZUO     NYSE                    Zuora, Inc.       19.79   
6.819191e+08   ZUMZ   NASDAQ                    Zumiez Inc.       26.72   

              ADR TSO  IPO_Year             Sector  \
Market_Cap                                           
1.618349e+08      NaN       NaN        Health Care   
1.653995e+08      NaN    2015.0        Health Care   
5.040782e+08      NaN    2017.0        Health Care   
2.147266e+09      NaN    2018.0         Technology   
6.819191e+08      NaN    2005.0  Consumer Services   

                                                       Industry  
Market_Cap                                                       
1.618349e+08  Biotechnology: Electromedical & Electrotherape...  
1.653995e+08                              Major Pharmaceuticals  
5.040782e+08                              Major Pharmaceuticals  
2.147266e+09            Computer Software: Prepackaged Software  
6.819191e+08                     Clothing/Shoe/Accessory Stores

In [139]:
  # Change the index
listings.sort_index(ascending=False, inplace=True)
listings.head()

Symbol Exchange                   Name  Last_Price  ADR TSO  \
Market_Cap                                                                 
9.459795e+11   AAPL   NASDAQ             Apple Inc.      200.62      NaN   
9.221233e+11   MSFT   NASDAQ  Microsoft Corporation      120.19      NaN   
9.074138e+11   AMZN   NASDAQ       Amazon.com, Inc.     1847.33      NaN   
8.387076e+11  GOOGL   NASDAQ          Alphabet Inc.     1206.45      NaN   
8.357253e+11   GOOG   NASDAQ          Alphabet Inc.     1202.16      NaN   

              IPO_Year             Sector  \
Market_Cap                                  
9.459795e+11    1980.0         Technology   
9.221233e+11    1986.0         Technology   
9.074138e+11    1997.0  Consumer Services   
8.387076e+11       NaN         Technology   
8.357253e+11    2004.0         Technology   

                                                     Industry  
Market_Cap                                                     
9.459795e+11                           Computer Manufacturing  
9.221233e+11          Computer Software: Prepackaged Software  
9.074138e+11                   Catalog/Specialty Distribution  
8.387076e+11  Computer Software: Programming, Data Processing  
8.357253e+11  Computer Software: Programming, Data Processing

#### Not surprisingly, here as well, the thrid largest Market cap is AMZN.

45. __Select__ the __15__ listings with the __lowest Prices__ (Penny Stocks)! Which Company has the __fourth lowest__ Stock Price?

In [143]:
# First let reset the index

listings.reset_index(inplace=True)

In [149]:
# THEN,

listings.nsmallest(n=15, columns="Last_Price")

index  Market_Cap Symbol Exchange  \
6617   6617         NaN  CYTXZ   NASDAQ   
6408   6408         NaN  JASNW   NASDAQ   
5998   5998         NaN  VEACW   NASDAQ   
5971   5971         NaN  WHLRW   NASDAQ   
6114   6114         NaN  SNOAW   NASDAQ   
6364   6364         NaN  LMFAW   NASDAQ   
6779   6779         NaN  APDNW   NASDAQ   
6786   6786         NaN  AMRWW   NASDAQ   
6545   6545         NaN  EYEGW   NASDAQ   
6592   6592         NaN  DRIOW   NASDAQ   
6127   6127         NaN  SHIPW   NASDAQ   
6691   6691         NaN  CETXW   NASDAQ   
6270   6270         NaN  NUROW   NASDAQ   
6223   6223         NaN  PHIOW   NASDAQ   
6249   6249         NaN  OXBRW   NASDAQ   

                                            Name  Last_Price  ADR TSO  \
6617                    Cytori Therapeutics Inc.      0.0030      NaN   
6408                      Jason Industries, Inc.      0.0040      NaN   
5998            Vantage Energy Acquisition Corp.      0.0050      NaN   
5971  Wheeler Real Estate Investment Trust, Inc.      0.0070      NaN   
6114                Sonoma Pharmaceuticals, Inc.      0.0100      NaN   
6364                    LM Funding America, Inc.      0.0115      NaN   
6779                    Applied DNA Sciences Inc      0.0151      NaN   
6786                   Alta Mesa Resources, Inc.      0.0207      NaN   
6545               Eyegate Pharmaceuticals, Inc.      0.0213      NaN   
6592                           DarioHealth Corp.      0.0230      NaN   
6127             Seanergy Maritime Holdings Corp      0.0270      NaN   
6691                                Cemtrex Inc.      0.0300      NaN   
6270                           NeuroMetrix, Inc.      0.0354      NaN   
6223                  Phio Pharmaceuticals Corp.      0.0388      NaN   
6249                Oxbridge Re Holdings Limited      0.0401      NaN   

      IPO_Year             Sector                                Industry  
6617       NaN        Health Care              Medical/Dental Instruments  
6408    2013.0  Consumer Durables  Miscellaneous manufacturing industries  
5998    2017.0            Finance                       Business Services  
5971       NaN  Consumer Services           Real Estate Investment Trusts  
6114       NaN        Health Care                   Major Pharmaceuticals  
6364    2015.0            Finance              Finance: Consumer Services  
6779       NaN  Consumer Services                 Other Consumer Services  
6786    2017.0             Energy                    Oil & Gas Production  
6545       NaN        Health Care                   Major Pharmaceuticals  
6592       NaN        Health Care              Medical/Dental Instruments  
6127       NaN     Transportation                   Marine Transportation  
6691       NaN      Capital Goods         Industrial Machinery/Components  
6270       NaN        Health Care              Medical/Dental Instruments  
6223       NaN        Health Care                   Major Pharmaceuticals  
6249    2014.0            Finance              Property-Casualty Insurers

#### The answer is Wheeler Real estate at the price of 0.0070

46. __Check__ for all elements in the column __"Exchange"__, whether they are __equal to "NASDAQ"__. __Save__ the boolean Series in the variable __mask1__! <br>
__Fill in the gaps__!

In [153]:
mask1=listings.Exchange=="NASDAQ"
mask1

0        True
1        True
2        True
3        True
4        True
        ...  
6847     True
6848    False
6849    False
6850    False
6851     True
Name: Exchange, Length: 6852, dtype: bool

47. __Check__ for all elements in the column __"Last_Price"__, whether they are __less than 1.0__. __Save__ the boolean Series in the variable __mask2__!

In [155]:
mask2=listings.Last_Price<1.0
mask2

0       False
1       False
2       False
3       False
4       False
        ...  
6847    False
6848    False
6849    False
6850    False
6851    False
Name: Last_Price, Length: 6852, dtype: bool

48. __Filter__ the DataFrame __listings__ for all __NASDAQ-listings__ (mask1) and __save__ the resulting DataFrame in the variable __nasdaq__ (create a copy)! <br>
__Fill in the gaps__!

In [157]:
nasdaq=listings.loc[listings.Exchange=="NASDAQ"].copy()
nasdaq

index    Market_Cap Symbol Exchange                               Name  \
0         0  9.459795e+11   AAPL   NASDAQ                         Apple Inc.   
1         1  9.221233e+11   MSFT   NASDAQ              Microsoft Corporation   
2         2  9.074138e+11   AMZN   NASDAQ                   Amazon.com, Inc.   
3         3  8.387076e+11  GOOGL   NASDAQ                      Alphabet Inc.   
4         4  8.357253e+11   GOOG   NASDAQ                      Alphabet Inc.   
...     ...           ...    ...      ...                                ...   
6844   6844           NaN  ADILW   NASDAQ         Adial Pharmaceuticals, Inc   
6845   6845           NaN  ACGLP   NASDAQ            Arch Capital Group Ltd.   
6846   6846           NaN  ACGLO   NASDAQ            Arch Capital Group Ltd.   
6847   6847           NaN  ACAMU   NASDAQ  Acamar Partners Acquisition Corp.   
6851   6851           NaN  ABEOW   NASDAQ           Abeona Therapeutics Inc.   

      Last_Price  ADR TSO  IPO_Year             Sector  \
0       200.6200      NaN    1980.0         Technology   
1       120.1900      NaN    1986.0         Technology   
2      1847.3300      NaN    1997.0  Consumer Services   
3      1206.4500      NaN       NaN         Technology   
4      1202.1600      NaN    2004.0         Technology   
...          ...      ...       ...                ...   
6844      0.3705      NaN    2018.0        Health Care   
6845     23.5300      NaN       NaN            Finance   
6846     24.6000      NaN       NaN            Finance   
6847     10.0100      NaN    2019.0            Finance   
6851      3.0700      NaN       NaN        Health Care   

                                             Industry  
0                              Computer Manufacturing  
1             Computer Software: Prepackaged Software  
2                      Catalog/Specialty Distribution  
3     Computer Software: Programming, Data Processing  
4     Computer Software: Programming, Data Processing  
...                                               ...  
6844                            Major Pharmaceuticals  
6845                       Property-Casualty Insurers  
6846                       Property-Casualty Insurers  
6847                                Business Services  
6851                            Major Pharmaceuticals  

[3435 rows x 10 columns]

__How many NASDAQ-listings__ does the __nasdaq__ DataFrame contain?

In [159]:
nasdaq.describe()

index    Market_Cap   Last_Price       ADR TSO     IPO_Year
count  3435.000000  3.168000e+03  3432.000000  1.400000e+02  1599.000000
mean   3687.723726  4.395262e+09    32.302384  5.869242e+07  2009.221388
std    1739.215970  3.831888e+10    72.124304  1.629909e+08     9.909466
min       0.000000  0.000000e+00     0.003000  2.400000e+04  1972.000000
25%    2313.000000  6.576675e+07     5.020000  4.271224e+06  2002.000000
50%    3874.000000  2.805001e+08    16.745000  1.406964e+07  2014.000000
75%    5141.500000  1.292507e+09    35.257500  3.962613e+07  2017.000000
max    6851.000000  9.459795e+11  1847.330000  1.485426e+09  2019.000000

#### The answer is 3435

49. __Filter__ the DataFrame __listings__ for all __Penny Stocks__ (mask2) and __save__ the resulting DataFrame in the variable __penny__ (create a copy)!

In [160]:
penny=listings.loc[mask2].copy()
penny

index    Market_Cap Symbol Exchange                              Name  \
2892   2892  6.130630e+08    RAD     NYSE              Rite Aid Corporation   
2906   2906  6.034052e+08    WFT     NYSE     Weatherford International plc   
2910   2910  6.009313e+08   GSAT     AMEX                  Globalstar, Inc.   
3017   3017  5.386351e+08    NGD     AMEX                     New Gold Inc.   
3483   3483  3.381600e+08    DNN     AMEX                 Denison Mine Corp   
...     ...           ...    ...      ...                               ...   
6808   6808           NaN  ALGRR   NASDAQ              Allegro Merger Corp.   
6810   6810           NaN  ALACW   NASDAQ  Alberton Acquisition Corporation   
6812   6812           NaN  ALACR   NASDAQ  Alberton Acquisition Corporation   
6835   6835           NaN  AGFSW   NASDAQ         AgroFresh Solutions, Inc.   
6844   6844           NaN  ADILW   NASDAQ        Adial Pharmaceuticals, Inc   

      Last_Price  ADR TSO  IPO_Year             Sector  \
2892      0.5677      NaN       NaN        Health Care   
2906      0.6014      NaN       NaN             Energy   
2910      0.4150      NaN    2006.0  Consumer Services   
3017      0.9301      NaN       NaN   Basic Industries   
3483      0.5740      NaN       NaN   Basic Industries   
...          ...      ...       ...                ...   
6808      0.2700      NaN    2018.0            Finance   
6810      0.0800      NaN    2018.0            Finance   
6812      0.2200      NaN    2018.0            Finance   
6835      0.0600      NaN    2014.0            Finance   
6844      0.3705      NaN    2018.0        Health Care   

                          Industry  
2892      Medical/Nursing Services  
2906            Metal Fabrications  
2910  Telecommunications Equipment  
3017               Precious Metals  
3483               Precious Metals  
...                            ...  
6808             Business Services  
6810             Business Services  
6812             Business Services  
6835             Business Services  
6844         Major Pharmaceuticals  

[327 rows x 10 columns]

__How many Penny Stocks__ does the __penny__ DataFrame contain?

#### It consists of 327 stocks

50. __Filter__ the DataFrame __listings__ for all __Penny Stocks__ (mask2) that are __NASDAQ-listed__ (mask1) and __save__ the resulting DataFrame in the variable __nasdaq_penny__! __Fill in the gaps!__

In [162]:
nasdaq_penny=listings.loc[mask1 & mask2]
nasdaq_penny

index    Market_Cap Symbol Exchange                              Name  \
3711   3711  2.679643e+08   NVAX   NASDAQ                     Novavax, Inc.   
4156   4156  1.768699e+08   MPVD   NASDAQ   Mountain Province Diamonds Inc.   
4526   4526  1.107320e+08    UPL   NASDAQ             Ultra Petroleum Corp.   
4535   4535  1.099413e+08   DRRX   NASDAQ                DURECT Corporation   
4553   4553  1.072175e+08    AMR   NASDAQ         Alta Mesa Resources, Inc.   
...     ...           ...    ...      ...                               ...   
6808   6808           NaN  ALGRR   NASDAQ              Allegro Merger Corp.   
6810   6810           NaN  ALACW   NASDAQ  Alberton Acquisition Corporation   
6812   6812           NaN  ALACR   NASDAQ  Alberton Acquisition Corporation   
6835   6835           NaN  AGFSW   NASDAQ         AgroFresh Solutions, Inc.   
6844   6844           NaN  ADILW   NASDAQ        Adial Pharmaceuticals, Inc   

      Last_Price  ADR TSO  IPO_Year            Sector  \
3711      0.5708      NaN       NaN       Health Care   
4156      0.8418      NaN       NaN  Basic Industries   
4526      0.5610      NaN       NaN            Energy   
4535      0.6784      NaN    2000.0       Health Care   
4553      0.2815      NaN    2017.0            Energy   
...          ...      ...       ...               ...   
6808      0.2700      NaN    2018.0           Finance   
6810      0.0800      NaN    2018.0           Finance   
6812      0.2200      NaN    2018.0           Finance   
6835      0.0600      NaN    2014.0           Finance   
6844      0.3705      NaN    2018.0       Health Care   

                                               Industry  
3711  Biotechnology: Biological Products (No Diagnos...  
4156                                    Precious Metals  
4526                               Oil & Gas Production  
4535                              Major Pharmaceuticals  
4553                               Oil & Gas Production  
...                                                 ...  
6808                                  Business Services  
6810                                  Business Services  
6812                                  Business Services  
6835                                  Business Services  
6844                              Major Pharmaceuticals  

[257 rows x 10 columns]

__How many listings__ does the __nasdaq_penny__ DataFrame contain?

#### There are 257 listings on the nasdaq_penny df

#### Intro to NA Values